gateway/src/metrics.rs

```diff
@@ -0,0 +1,1177 @@
+                        );
+                    } else {
+                        slog::error!(
+                            &self.log,
+                            "a spawned SP poller task panicked! this should \
+                             never happen: we compile with `panic=\"abort\"`, so \
+                             a spawned task panicking should abort the whole \
+                             process...";
+                            "error" => %e,
+                        );
+                    }
+                }
+                // drain any remaining errors
+                joined = self.tasks.try_join_next();
+            }
+
+            for (spid, poller) in &mut known_sps {
+                // There's still someone checking up on this SP, right?
+                if poller.is_finished() {
+                    // Welp.
+                    slog::info!(
+                        &self.log,
+                        "uh-oh! a known SP's poller task has gone AWOL. restarting it...";
+                        "sp_slot" => ?spid.slot,
+                        "chassis_type" => ?spid.typ,
+                    );
+                    *poller =
+                        self.tasks.spawn(self.make_poller(*spid, rack_id));
+                }
+            }
+        }
+    }
+
+    fn make_poller(
+        &self,
+        spid: SpIdentifier,
+        rack_id: Uuid,
+    ) -> impl std::future::Future<Output = Result<(), SpCommsError>> {
+        let poller = SpPoller {
+            spid,
+            rack_id,
+            mgs_id: self.mgs_id,
+            log: self.log.new(slog::o!(
+                "sp_slot" => spid.slot,
+                "chassis_type" => format!("{:?}", spid.typ),
+            )),
+            components: HashMap::new(),
+            known_state: None,
+            sample_tx: self.sample_tx.clone(),
+        };
+        poller.run(self.poll_interval, self.apictx.clone())
+    }
+}
+
+impl Drop for PollerManager {
+    fn drop(&mut self) {
+        // This is why the `JoinSet` is a field on the `PollerManager` struct
+        // rather than a local variable in `async fn run()`!
+        self.tasks.abort_all();
+    }
+}
+
+impl SpPoller {
+    async fn run(
+        mut self,
+        poll_interval: Duration,
+        apictx: Arc<ServerContext>,
+    ) -> Result<(), SpCommsError> {
+        let mut interval = tokio::time::interval(poll_interval);
+        let switch = &apictx.mgmt_switch;
+        let sp = switch.sp(self.spid)?;
+        loop {
+            interval.tick().await;
+            slog::trace!(&self.log, "interval elapsed, polling SP...");
+
+            match self.poll(sp).await {
+                // No sense cluttering the ringbuffer with empty vecs...
+                Ok(samples) if samples.is_empty() => {
+                    slog::trace!(
+                        &self.log,
+                        "polled SP, no samples returned";
+                        "num_samples" => 0usize
```

Comment:
The `num_samples` key seems implied by the message, I think?

Reply:
I included it because I wanted to be able to use `looker` or `grep` to see _all_ log entries that had `num_samples` in them.